File: 07_midas_aruco_fusion/core/gui_fusion.py
import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, GdkPixbuf, GLib, Gdk
import cv2
import queue

class FusionGUI(Gtk.Window):
    def __init__(self, moil_undistorter=None, headless=False, initial_exposure=0,
                 exposure_callback=None, gain_callback=None, brightness_callback=None):
        super().__init__(title="ArUco + MiDaS | Fusion Interface")
        self.set_default_size(1280, 720)
        
        self.moil_undistorter = moil_undistorter
        self.headless = headless
        self.smart_exposure_callback = exposure_callback
        self.initial_exposure = initial_exposure
        self._alive = True
        self._last_ui_frame_t = 0.0
        self.calibration_ready_event = __import__('threading').Event()
        # Cache boolean untuk normalize lighting — dibaca dari background thread
        # WAJIB Python bool biasa, bukan akses GTK widget (thread-unsafe → SIGABRT)
        self._normalize_enabled = False
        self._bw_enabled = False
        self.key_queue = queue.Queue()
        
        self.connect("destroy", self.on_destroy)
        self.setup_ui()
        
    def setup_ui(self):
        hbox = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=10)
        self.add(hbox)
        
        # ─── Sidebar Kiri (Controls) ───
        vbox_ctrl = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=5)
        vbox_ctrl.set_border_width(10)
        vbox_ctrl.set_size_request(320, -1)
        
        # System Status Frame
        f_sys = Gtk.Frame(label="System Status")
        vb_s = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=5)
        vb_s.set_border_width(5)
        self.lbl_status_calib = Gtk.Label(label="Mode: Live")
        self.lbl_status_ai = Gtk.Label(label="AI: Ready")
        vb_s.pack_start(self.lbl_status_calib, False, False, 0)
        vb_s.pack_start(self.lbl_status_ai, False, False, 0)
        
        # Smart Exposure Control
        # Menggabungkan Exposure (waktu), Gain (ISO), dan Brightness (EV) ke dalam satu slider 1.0 - 10.0
        hb_exp = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=3)
        hb_exp.pack_start(Gtk.Label(label="Smart Exp:"), False, False, 0)
        
        btn_exp_mm = Gtk.Button(label="--")
        btn_exp_mm.connect("clicked", self.on_adj_smart_exposure, -2.0)
        btn_exp_min = Gtk.Button(label=" - ")
        btn_exp_min.connect("clicked", self.on_adj_smart_exposure, -0.5)
        btn_exp_plus = Gtk.Button(label=" + ")
        btn_exp_plus.connect("clicked", self.on_adj_smart_exposure, +0.5)
        btn_exp_pp = Gtk.Button(label="++")
        btn_exp_pp.connect("clicked", self.on_adj_smart_exposure, +2.0)
        
        display_val = max(1.0, min(10.0, round(self.initial_exposure / 1000, 1)))
        self.entry_exposure = Gtk.Entry(text=str(display_val))
        self.entry_exposure.set_tooltip_text("1.0 (Gelap) - 10.0 (Sangat Terang). Otomatis mengatur Shutter, ISO & EV.")
        self.entry_exposure.set_width_chars(6)
        
        hb_exp.pack_start(btn_exp_mm, False, False, 0)
        hb_exp.pack_start(btn_exp_min, False, False, 0)
        hb_exp.pack_start(self.entry_exposure, True, True, 0)
        hb_exp.pack_start(btn_exp_plus, False, False, 0)
        hb_exp.pack_start(btn_exp_pp, False, False, 0)
        
        btn_apply_exp = Gtk.Button(label="Apply")
        btn_apply_exp.connect("clicked", self.on_apply_smart_exposure)
        hb_exp.pack_start(btn_apply_exp, False, False, 0)
        
        vb_s.pack_start(hb_exp, False, False, 5)
        
        f_sys.add(vb_s)
        vbox_ctrl.pack_start(f_sys, False, False, 5)
        
        # Moildev Anypoint Frame
        f_moil = Gtk.Frame(label="Moildev Anypoint Controls")
        vb_m = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=5)
        vb_m.set_border_width(5)
        
        # Alpha (Pitch)
        hb_a = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=5)
        hb_a.pack_start(Gtk.Label(label="Alpha (Pitch):"), False, False, 0)
        btn_a_min = Gtk.Button(label=" - ")
        btn_a_min.connect("clicked", self.on_adj_alpha, -5.0)
        btn_a_plus = Gtk.Button(label=" + ")
        btn_a_plus.connect("clicked", self.on_adj_alpha, +5.0)
        self.entry_alpha = Gtk.Entry(text="0.0")
        hb_a.pack_start(btn_a_min, False, False, 0)
        hb_a.pack_start(self.entry_alpha, True, True, 0)
        hb_a.pack_start(btn_a_plus, False, False, 0)
        vb_m.pack_start(hb_a, False, False, 0)
        
        # Beta (Yaw)
        hb_b = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=5)
        hb_b.pack_start(Gtk.Label(label="Beta (Yaw):  "), False, False, 0)
        btn_b_min = Gtk.Button(label=" - ")
        btn_b_min.connect("clicked", self.on_adj_beta, -5.0)
        btn_b_plus = Gtk.Button(label=" + ")
        btn_b_plus.connect("clicked", self.on_adj_beta, +5.0)
        self.entry_beta = Gtk.Entry(text="0.0")
        hb_b.pack_start(btn_b_min, False, False, 0)
        hb_b.pack_start(self.entry_beta, True, True, 0)
        hb_b.pack_start(btn_b_plus, False, False, 0)
        vb_m.pack_start(hb_b, False, False, 0)
        
        # Zoom
        hb_z = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=5)
        hb_z.pack_start(Gtk.Label(label="Zoom:          "), False, False, 0)
        btn_z_min = Gtk.Button(label=" - ")
        btn_z_min.connect("clicked", self.on_adj_zoom, -0.1)
        btn_z_plus = Gtk.Button(label=" + ")
        btn_z_plus.connect("clicked", self.on_adj_zoom, +0.1)
        self.entry_zoom = Gtk.Entry(text="1.4")
        hb_z.pack_start(btn_z_min, False, False, 0)
        hb_z.pack_start(self.entry_zoom, True, True, 0)
        hb_z.pack_start(btn_z_plus, False, False, 0)
        vb_m.pack_start(hb_z, False, False, 0)
        
        # Apply & Reset Buttons
        hb_moil_btns = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=5)
        btn_apply = Gtk.Button(label="Apply Anypoint")
        btn_apply.connect("clicked", self.on_apply_anypoint)
        btn_reset = Gtk.Button(label="Reset View")
        btn_reset.connect("clicked", self.on_reset_anypoint)
        hb_moil_btns.pack_start(btn_apply, True, True, 0)
        hb_moil_btns.pack_start(btn_reset, True, True, 0)
        vb_m.pack_start(hb_moil_btns, False, False, 5)
        
        f_moil.add(vb_m)
        vbox_ctrl.pack_start(f_moil, False, False, 5)
        
        # Actions Frame
        f_act = Gtk.Frame(label="Actions")
        vb_a = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=5)
        vb_a.set_border_width(5)
        
        # Tombol Start Calibration — hanya aktif di setup mode sebelum kalibrasi
        self.btn_start_calib = Gtk.Button(label="▶  Start Calibration")
        self.btn_start_calib.connect("clicked", self.on_start_calibration)
        self.btn_start_calib.get_style_context().add_class("suggested-action")
        vb_a.pack_start(self.btn_start_calib, False, False, 0)

        # Tombol Next Step (Pengganti Spasi)
        self.btn_next_step = Gtk.Button(label="⏭  Next Step (Space)")
        self.btn_next_step.connect("clicked", lambda w: self.queue_key(ord(' ')))
        self.btn_next_step.get_style_context().add_class("suggested-action")
        vb_a.pack_start(self.btn_next_step, False, False, 0)
        
        # Toggle Normalize Lighting
        self.chk_normalize = Gtk.CheckButton(label="Enable Normalize Lighting")
        self.chk_normalize.set_active(False)  # default mati
        # Update cache Python saat checkbox diubah (JANGAN baca widget dari background thread!)
        self.chk_normalize.connect("toggled", self._on_normalize_toggled)
        vb_a.pack_start(self.chk_normalize, False, False, 5)
        
        # Toggle Black & White Mode
        self.chk_bw = Gtk.CheckButton(label="Enable Black & White Mode")
        self.chk_bw.set_active(False)
        self.chk_bw.connect("toggled", self._on_bw_toggled)
        vb_a.pack_start(self.chk_bw, False, False, 5)
        
        self.lbl_setup_hint = Gtk.Label(label="")
        self.lbl_setup_hint.set_line_wrap(True)
        vb_a.pack_start(self.lbl_setup_hint, False, False, 0)
        
        btn_record = Gtk.Button(label="Toggle Record (R)")
        btn_record.connect("clicked", lambda w: self.queue_key(ord('r')))
        
        btn_screenshot = Gtk.Button(label="Screenshot (S)")
        btn_screenshot.connect("clicked", lambda w: self.queue_key(ord('s')))
        
        btn_cap = Gtk.Button(label="Capture Calib Point (C)")
        btn_cap.connect("clicked", lambda w: self.queue_key(ord('c')))
        
        btn_quit = Gtk.Button(label="Quit (ESC)")
        btn_quit.connect("clicked", lambda w: self.queue_key(27))
        
        vb_a.pack_start(btn_record, False, False, 0)
        vb_a.pack_start(btn_screenshot, False, False, 0)
        vb_a.pack_start(btn_cap, False, False, 0)
        vb_a.pack_start(btn_quit, False, False, 0)
        
        # Sembunyikan tombol Start Calibration & Next Step secara default
        self.btn_start_calib.hide()
        self.btn_next_step.hide()
        self.lbl_setup_hint.hide()
        
        f_act.add(vb_a)
        vbox_ctrl.pack_start(f_act, False, False, 5)
        
        hbox.pack_start(vbox_ctrl, False, False, 0)
        
        # ─── Main Video Feed (Kanan) ───
        self.image = Gtk.Image()
        hbox.pack_start(self.image, True, True, 0)
        
        # Listen for key presses on the window itself
        self.connect("key-press-event", self.on_key_press)
        
        # Inisialisasi nilai awal
        if self.moil_undistorter:
            self.entry_alpha.set_text(str(self.moil_undistorter.pitch))
            self.entry_beta.set_text(str(self.moil_undistorter.yaw))
            self.entry_zoom.set_text(str(self.moil_undistorter.zoom))

    def update_image(self, frame_bgr):
        """Called by background thread. Throttled to max 20 FPS to avoid GLib queue flooding."""
        if self.headless or not self._alive:
            return

        import time
        now = time.monotonic()
        if now - self._last_ui_frame_t < 0.05:  # max 20 FPS
            return
        self._last_ui_frame_t = now

        # Lakukan resize dan konversi BGR→RGB di background thread (aman, tidak menyentuh GTK)
        h, w = frame_bgr.shape[:2]
        target_w = 960
        if w > target_w:
            scale = target_w / float(w)
            new_h = int(h * scale)
            # Menggunakan INTER_AREA sangat penting untuk downscaling resolusi tinggi agar tidak ada aliasing/garis gergaji di layar
            frame_bgr = cv2.resize(frame_bgr, (target_w, new_h), interpolation=cv2.INTER_AREA)

        rgb = cv2.cvtColor(frame_bgr, cv2.COLOR_BGR2RGB)
        out_h, out_w = rgb.shape[:2]
        # Salin data ke GLib.Bytes — dikelola oleh GLib (aman lintas thread)
        glib_bytes = GLib.Bytes.new(rgb.tobytes())
        GLib.idle_add(self._update_image_from_glib_bytes, glib_bytes, out_w, out_h)

    def _update_image_from_glib_bytes(self, glib_bytes, w, h):
        """Dijalankan di GTK main thread — aman memanggil widget GTK.
        WAJIB pakai new_from_bytes (bukan new_from_data): new_from_data tidak menyalin
        data sehingga saat Python GC membebaskan glib_bytes, GTK akan akses dangling pointer.
        new_from_bytes menyalin data ke memory yang dikelola GLib — aman dari GC Python.
        """
        if self.headless or not self._alive:
            return False
        try:
            pb = GdkPixbuf.Pixbuf.new_from_bytes(
                glib_bytes,
                GdkPixbuf.Colorspace.RGB,
                False,   # has_alpha
                8,       # bits_per_sample
                w, h,
                w * 3,   # rowstride
            )
            self.image.set_from_pixbuf(pb)
        except Exception:
            pass
        return False

    def on_adj_alpha(self, widget, delta):
        try:
            val = float(self.entry_alpha.get_text()) + delta
            self.entry_alpha.set_text(str(round(val, 1)))
            self.on_apply_anypoint(None)
        except ValueError:
            pass

    def on_adj_beta(self, widget, delta):
        try:
            val = float(self.entry_beta.get_text()) + delta
            self.entry_beta.set_text(str(round(val, 1)))
            self.on_apply_anypoint(None)
        except ValueError:
            pass

    def on_adj_zoom(self, widget, delta):
        try:
            val = float(self.entry_zoom.get_text()) + delta
            self.entry_zoom.set_text(str(round(val, 2)))
            self.on_apply_anypoint(None)
        except ValueError:
            pass
            
    def on_adj_smart_exposure(self, widget, delta):
        try:
            val = round(float(self.entry_exposure.get_text()) + delta, 1)
            val = max(1.0, min(10.0, val))
            self.entry_exposure.set_text(str(val))
            self.on_apply_smart_exposure(None)
        except ValueError:
            pass

    def on_apply_smart_exposure(self, widget):
        if self.smart_exposure_callback:
            try:
                val = float(self.entry_exposure.get_text())
                val = max(1.0, min(10.0, val))
                
                # Pemetaan Smart Exposure (1.0 - 10.0):
                # Exposure time (Raw): 1000 -> 10000
                raw_exp = int(round(val * 1000))
                
                # Gain (0 - 255): Scale linearly from val 1.0 to 10.0
                raw_gain = int(round((val - 1.0) / 9.0 * 255.0))
                
                # Brightness EV (-64 to +64): Scale linearly
                raw_bri = int(round((val - 1.0) / 9.0 * 128.0 - 64.0))
                
                # Panggil satu callback dengan 3 nilai
                self.smart_exposure_callback(raw_exp, raw_gain, raw_bri)
            except ValueError:
                pass

    def on_apply_anypoint(self, widget):
        if not self.moil_undistorter:
            return
        try:
            a = float(self.entry_alpha.get_text())
            b = float(self.entry_beta.get_text())
            z = float(self.entry_zoom.get_text())
            # KRITIS: Harus memanggil update_maps() agar Moildev benar-benar
            # meregenerasi remap matrices. Sebelumnya hanya meng-set atribut Python
            # tanpa regenerasi maps → zoom/alpha/beta tidak berpengaruh secara visual.
            self.moil_undistorter.update_maps(pitch=a, yaw=b, zoom=z)
            # Sinkronkan entry dengan nilai aktual (hybrid zoom mungkin mengubah display)
            self.entry_zoom.set_text(str(round(self.moil_undistorter.zoom, 2)))
        except ValueError:
            pass

    def on_reset_anypoint(self, widget):
        self.entry_alpha.set_text("0.0")
        self.entry_beta.set_text("0.0")
        self.entry_zoom.set_text("1.4")
        if self.moil_undistorter:
            self.moil_undistorter.update_maps(pitch=0.0, yaw=0.0, roll=0.0, zoom=1.4)
            self.entry_zoom.set_text(str(round(self.moil_undistorter.zoom, 2)))

    def on_start_calibration(self, widget):
        """Dipanggil saat user menekan tombol Start Calibration."""
        self.calibration_ready_event.set()
        GLib.idle_add(lambda: (self.btn_start_calib.set_sensitive(False),
                               self.btn_start_calib.hide(),
                               self.btn_next_step.show(),
                               self.lbl_setup_hint.set_text("Calibration in progress...\nClick 'Next Step' when prompted.")) or False)

    def enter_setup_mode(self, calib_name="Calibration"):
        """Dipanggil dari worker thread untuk menampilkan tombol Start Calibration."""
        hint = f"Setup mode: adjust exposure & anypoint,\nthen click \u25b6 Start {calib_name}"
        GLib.idle_add(lambda: (self.btn_start_calib.show(),
                               self.lbl_setup_hint.show(),
                               self.lbl_setup_hint.set_text(hint),
                               self.lbl_status_calib.set_text(f"Setup: {calib_name}")) or False)

    def _on_normalize_toggled(self, widget):
        """Dijalankan di GTK main thread — update cache Python yang aman dibaca dari thread lain."""
        self._normalize_enabled = widget.get_active()

    def is_normalize_enabled(self):
        """Dibaca dari background thread — hanya kembalikan Python bool (BUKAN akses GTK widget)."""
        return self._normalize_enabled

    def _on_bw_toggled(self, widget):
        """Dijalankan di GTK main thread — update cache Python yang aman dibaca dari thread lain."""
        self._bw_enabled = widget.get_active()

    def is_bw_enabled(self):
        """Dibaca dari background thread — hanya kembalikan Python bool."""
        return self._bw_enabled

    def on_key_press(self, widget, event):
        # Convert GDK keyval to ascii
        keyname = Gdk.keyval_name(event.keyval)
        if keyname and len(keyname) == 1:
            self.queue_key(ord(keyname.lower()))
        elif keyname == "Escape":
            self.queue_key(27)
        return False

    def queue_key(self, key_code):
        self.key_queue.put(key_code)

    def get_key(self):
        """Pengganti cv2.waitKey(1) untuk worker thread."""
        try:
            return self.key_queue.get_nowait()
        except queue.Empty:
            return -1

    def on_destroy(self, widget):
        self._alive = False   # Hentikan semua idle callbacks sebelum Gtk.main_quit
        self.queue_key(27)    # Kirim ESC ke background thread agar loop berhenti
        Gtk.main_quit()

File: 08_midas_aruco_fusion_cpp/src/gui/gui_fusion.cpp
/**
 * @file gui_fusion.cpp
 * @brief GTKmm-3.0 GUI implementation.
 * Identical layout to Python 07_midas_aruco_fusion.
 */

#ifdef HAS_GTKMM

#include "gui_fusion.h"
#include <opencv2/imgproc.hpp>
#include <sstream>
#include <iomanip>
#include <iostream>

namespace fusion {

FusionGUI::FusionGUI() {
    set_title("ArUco + MiDaS | Fusion Interface (C++)");
    set_default_size(1280, 720);
    
    // Enable dark theme to match typical Python GTK feel (optional but nice)
    auto settings = Gtk::Settings::get_default();
    if (settings) {
        settings->property_gtk_application_prefer_dark_theme() = true;
    }

    setup_ui();
    connect_signals();
}

void FusionGUI::setup_ui() {
    main_box_.set_spacing(10);
    add(main_box_);

    // ─── Sidebar Kiri (Controls) ───
    left_panel_.set_spacing(5);
    left_panel_.set_border_width(10);
    left_panel_.set_size_request(320, -1);

    // System Status Frame
    auto f_sys = Gtk::make_managed<Gtk::Frame>("System Status");
    auto vb_s = Gtk::make_managed<Gtk::Box>(Gtk::ORIENTATION_VERTICAL, 5);
    vb_s->set_border_width(5);
    vb_s->pack_start(lbl_status_calib_, Gtk::PACK_SHRINK);
    vb_s->pack_start(lbl_status_ai_, Gtk::PACK_SHRINK);

    // Smart Exposure Control
    auto hb_exp = Gtk::make_managed<Gtk::Box>(Gtk::ORIENTATION_HORIZONTAL, 3);
    auto lbl_exp = Gtk::make_managed<Gtk::Label>("Smart Exp:");
    hb_exp->pack_start(*lbl_exp, Gtk::PACK_SHRINK);
    
    entry_exposure_.set_text("1.0");
    entry_exposure_.set_tooltip_text("1.0 (Gelap) - 10.0 (Sangat Terang). Otomatis mengatur Shutter, ISO & EV.");
    entry_exposure_.set_width_chars(6);

    hb_exp->pack_start(btn_exp_mm_, Gtk::PACK_SHRINK);
    hb_exp->pack_start(btn_exp_m_, Gtk::PACK_SHRINK);
    hb_exp->pack_start(entry_exposure_, Gtk::PACK_EXPAND_WIDGET);
    hb_exp->pack_start(btn_exp_p_, Gtk::PACK_SHRINK);
    hb_exp->pack_start(btn_exp_pp_, Gtk::PACK_SHRINK);
    hb_exp->pack_start(btn_apply_exp_, Gtk::PACK_SHRINK);
    
    vb_s->pack_start(*hb_exp, Gtk::PACK_SHRINK, 5);
    f_sys->add(*vb_s);
    left_panel_.pack_start(*f_sys, Gtk::PACK_SHRINK, 5);

    // Moildev Anypoint Frame
    auto f_moil = Gtk::make_managed<Gtk::Frame>("Moildev Anypoint Controls");
    auto vb_m = Gtk::make_managed<Gtk::Box>(Gtk::ORIENTATION_VERTICAL, 5);
    vb_m->set_border_width(5);

    auto make_adj_row = [](const std::string& label, Gtk::Button& b_min, Gtk::Entry& entry, Gtk::Button& b_plus) {
        auto hb = Gtk::make_managed<Gtk::Box>(Gtk::ORIENTATION_HORIZONTAL, 5);
        auto lbl = Gtk::make_managed<Gtk::Label>(label);
        lbl->set_width_chars(12);
        lbl->set_alignment(0.0, 0.5);
        hb->pack_start(*lbl, Gtk::PACK_SHRINK);
        
        b_min.set_label(" - ");
        b_plus.set_label(" + ");
        hb->pack_start(b_min, Gtk::PACK_SHRINK);
        hb->pack_start(entry, Gtk::PACK_EXPAND_WIDGET);
        hb->pack_start(b_plus, Gtk::PACK_SHRINK);
        return hb;
    };

    entry_alpha_.set_text("0.0");
    entry_beta_.set_text("0.0");
    entry_zoom_.set_text("1.4");

    auto btn_a_min = Gtk::make_managed<Gtk::Button>();
    auto btn_a_plus = Gtk::make_managed<Gtk::Button>();
    vb_m->pack_start(*make_adj_row("Alpha (Pitch):", *btn_a_min, entry_alpha_, *btn_a_plus), Gtk::PACK_SHRINK);
    
    auto btn_b_min = Gtk::make_managed<Gtk::Button>();
    auto btn_b_plus = Gtk::make_managed<Gtk::Button>();
    vb_m->pack_start(*make_adj_row("Beta (Yaw):", *btn_b_min, entry_beta_, *btn_b_plus), Gtk::PACK_SHRINK);
    
    auto btn_z_min = Gtk::make_managed<Gtk::Button>();
    auto btn_z_plus = Gtk::make_managed<Gtk::Button>();
    vb_m->pack_start(*make_adj_row("Zoom:", *btn_z_min, entry_zoom_, *btn_z_plus), Gtk::PACK_SHRINK);

    // Adj handlers
    btn_a_min->signal_clicked().connect([this]() { adj_alpha(-5.0); });
    btn_a_plus->signal_clicked().connect([this]() { adj_alpha(5.0); });
    btn_b_min->signal_clicked().connect([this]() { adj_beta(-5.0); });
    btn_b_plus->signal_clicked().connect([this]() { adj_beta(5.0); });
    btn_z_min->signal_clicked().connect([this]() { adj_zoom(-0.1); });
    btn_z_plus->signal_clicked().connect([this]() { adj_zoom(0.1); });

    auto hb_moil_btns = Gtk::make_managed<Gtk::Box>(Gtk::ORIENTATION_HORIZONTAL, 5);
    hb_moil_btns->pack_start(btn_apply_anypoint_, Gtk::PACK_EXPAND_WIDGET);
    hb_moil_btns->pack_start(btn_reset_anypoint_, Gtk::PACK_EXPAND_WIDGET);
    vb_m->pack_start(*hb_moil_btns, Gtk::PACK_SHRINK, 5);

    f_moil->add(*vb_m);
    left_panel_.pack_start(*f_moil, Gtk::PACK_SHRINK, 5);

    // Actions Frame
    auto f_act = Gtk::make_managed<Gtk::Frame>("Actions");
    auto vb_a = Gtk::make_managed<Gtk::Box>(Gtk::ORIENTATION_VERTICAL, 5);
    vb_a->set_border_width(5);

    vb_a->pack_start(btn_start_calib_, Gtk::PACK_SHRINK);
    vb_a->pack_start(btn_next_step_, Gtk::PACK_SHRINK);
    
    // Hide calib buttons by default
    btn_start_calib_.hide();
    btn_next_step_.hide();

    vb_a->pack_start(chk_normalize_, Gtk::PACK_SHRINK, 5);
    vb_a->pack_start(chk_bw_, Gtk::PACK_SHRINK, 5);
    vb_a->pack_start(chk_depth_, Gtk::PACK_SHRINK, 5);
    
    lbl_setup_hint_.set_line_wrap(true);
    vb_a->pack_start(lbl_setup_hint_, Gtk::PACK_SHRINK);

    vb_a->pack_start(btn_record_, Gtk::PACK_SHRINK);
    vb_a->pack_start(btn_snapshot_, Gtk::PACK_SHRINK);
    vb_a->pack_start(btn_cap_, Gtk::PACK_SHRINK);
    vb_a->pack_start(btn_quit_, Gtk::PACK_SHRINK);

    f_act->add(*vb_a);
    left_panel_.pack_start(*f_act, Gtk::PACK_SHRINK, 5);

    main_box_.pack_start(left_panel_, Gtk::PACK_SHRINK);

    // ─── Main Video Feed (Kanan) ───
    main_box_.pack_start(camera_view_, Gtk::PACK_EXPAND_WIDGET);

    show_all_children();
    btn_start_calib_.hide();
    btn_next_step_.hide();
}

void FusionGUI::adj_exposure(double delta) {
    try {
        double val = std::stod(entry_exposure_.get_text());
        val = std::max(1.0, std::min(10.0, val + delta));
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(1) << val;
        entry_exposure_.set_text(ss.str());
    } catch (...) {}
}

void FusionGUI::adj_alpha(double delta) {
    try {
        double val = std::stod(entry_alpha_.get_text());
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(1) << (val + delta);
        entry_alpha_.set_text(ss.str());
    } catch (...) {}
}

void FusionGUI::adj_beta(double delta) {
    try {
        double val = std::stod(entry_beta_.get_text());
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(1) << (val + delta);
        entry_beta_.set_text(ss.str());
    } catch (...) {}
}

void FusionGUI::adj_zoom(double delta) {
    try {
        double val = std::stod(entry_zoom_.get_text());
        val = std::max(0.1, val + delta);
        std::ostringstream ss;
        ss << std::fixed << std::setprecision(2) << val;
        entry_zoom_.set_text(ss.str());
    } catch (...) {}
}

void FusionGUI::connect_signals() {
    btn_exp_mm_.signal_clicked().connect([this]() { adj_exposure(-2.0); });
    btn_exp_m_.signal_clicked().connect([this]()  { adj_exposure(-0.5); });
    btn_exp_p_.signal_clicked().connect([this]()  { adj_exposure(0.5); });
    btn_exp_pp_.signal_clicked().connect([this]() { adj_exposure(2.0); });

    btn_apply_exp_.signal_clicked().connect([this]() {
        if (on_smart_exposure) {
            try { on_smart_exposure(std::stod(entry_exposure_.get_text())); } catch(...) {}
        }
    });

    btn_apply_anypoint_.signal_clicked().connect([this]() {
        if (on_apply_anypoint) {
            try {
                double a = std::stod(entry_alpha_.get_text());
                double b = std::stod(entry_beta_.get_text());
                double z = std::stod(entry_zoom_.get_text());
                on_apply_anypoint(a, b, z);
            } catch (...) {}
        }
    });

    btn_reset_anypoint_.signal_clicked().connect([this]() {
        entry_alpha_.set_text("0.0");
        entry_beta_.set_text("0.0");
        entry_zoom_.set_text("1.4");
        if (on_reset_anypoint) on_reset_anypoint();
    });

    btn_start_calib_.signal_clicked().connect([this]() { if (on_start_calib) on_start_calib(); });
    
    chk_normalize_.signal_toggled().connect([this]() {
        if (on_toggle_normalize) on_toggle_normalize(chk_normalize_.get_active());
    });
    
    chk_bw_.signal_toggled().connect([this]() {
        if (on_toggle_bw) on_toggle_bw(chk_bw_.get_active());
    });
    
    chk_depth_.signal_toggled().connect([this]() {
        if (on_toggle_depth) on_toggle_depth(chk_depth_.get_active());
    });

    btn_snapshot_.signal_clicked().connect([this]() { if (on_snapshot) on_snapshot(); });
    btn_quit_.signal_clicked().connect([this]() { close(); });

    btn_record_.signal_clicked().connect([this]() { if (on_queue_key) on_queue_key('r'); });
    btn_cap_.signal_clicked().connect([this]() { if (on_queue_key) on_queue_key('c'); });
    btn_next_step_.signal_clicked().connect([this]() { if (on_queue_key) on_queue_key(' '); });
}

Glib::RefPtr<Gdk::Pixbuf> FusionGUI::mat_to_pixbuf(const cv::Mat& frame) const {
    cv::Mat rgb;
    if (frame.channels() == 3) {
        cv::cvtColor(frame, rgb, cv::COLOR_BGR2RGB);
    } else if (frame.channels() == 1) {
        cv::cvtColor(frame, rgb, cv::COLOR_GRAY2RGB);
    } else {
        rgb = frame;
    }

    // Must clone data because Gdk::Pixbuf doesn't copy it when using create_from_data
    // For safety, we use deep copy to a vector or managed memory in production, 
    // but here we let OpenCV handle the Mat lifecycle since update_frame creates a local copy.
    // Wait, Gdk::Pixbuf requires data to stay alive. We must copy it.
    auto pb = Gdk::Pixbuf::create(Gdk::COLORSPACE_RGB, false, 8, rgb.cols, rgb.rows);
    
    int rowstride = pb->get_rowstride();
    guint8* pixels = pb->get_pixels();
    for (int y = 0; y < rgb.rows; ++y) {
        std::memcpy(pixels + y * rowstride, rgb.ptr(y), rgb.cols * 3);
    }
    
    return pb;
}

void FusionGUI::update_frame(const cv::Mat& frame) {
    if (frame.empty()) return;
    
    // Resize with INTER_AREA for high-quality downscale (similar to Python)
    cv::Mat display;
    int target_w = 960;
    if (frame.cols > target_w) {
        double scale = static_cast<double>(target_w) / frame.cols;
        int target_h = static_cast<int>(frame.rows * scale);
        cv::resize(frame, display, cv::Size(target_w, target_h), 0, 0, cv::INTER_AREA);
    } else {
        display = frame.clone();
    }
    
    auto pb = mat_to_pixbuf(display);
    camera_view_.set(pb);
}

void FusionGUI::set_status(const std::string& text) {
    lbl_status_ai_.set_text(text);
}

void FusionGUI::update_measurements(double height_cm, double distance_cm,
                                     double diameter_cm, double volume_ml) {
    (void)distance_cm;
    (void)diameter_cm;
    std::ostringstream ss;
    ss << std::fixed << std::setprecision(1);
    ss << "Height: " << height_cm << " cm | Vol: " << volume_ml << " mL";
    lbl_status_ai_.set_text(ss.str());
}

void FusionGUI::set_calib_mode(const std::string& mode) {
    lbl_status_calib_.set_text("Mode: " + mode);
}

bool FusionGUI::on_key_press_event(GdkEventKey* event) {
    if (event->keyval == GDK_KEY_q || event->keyval == GDK_KEY_Escape) {
        close();
        return true;
    }
    if (on_queue_key) {
        on_queue_key(event->keyval);
    }
    return Gtk::Window::on_key_press_event(event);
}

} // namespace fusion

#endif // HAS_GTKMM
